src/internal_telemetry/allocations/allocator/token.rs

```diff
@@ -0,0 +1,161 @@
+}
+
+/// The identifier that uniquely identifiers an allocation group.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct AllocationGroupId(NonZeroUsize);
+
+impl AllocationGroupId {
+    /// The group ID used for allocations which are not made within a registered allocation group.
+    pub const ROOT: Self = Self(unsafe { NonZeroUsize::new_unchecked(1) });
+
+    pub(super) const fn from_raw_unchecked(raw_group_id: usize) -> Self {
+        unsafe { Self(NonZeroUsize::new_unchecked(raw_group_id)) }
+    }
+
+    /// Gets the integer representation of this group ID.
+    #[must_use]
+    pub const fn as_raw(self) -> usize {
+        self.0.get()
+    }
+
+    /// Registers an allocation group ID.
+    ///
+    /// This group ID uniquely identifies a given allocation group, and is the means by which to
+    /// distinguish allocator events between various allocation groups.
+    ///
+    /// Group IDs must be attached to a [`Span`][tracing::Span] in order to become active,
+    /// assocating allocations and deallocations within an active span as being attached to the
+    /// given allocation group.
+    pub fn register() -> Option<AllocationGroupId> {
+        static GROUP_ID: AtomicUsize = AtomicUsize::new(AllocationGroupId::ROOT.0.get() + 1);
+        static HIGHEST_GROUP_ID: AtomicUsize =
+            AtomicUsize::new(AllocationGroupId::ROOT.0.get() + 1);
+
+        let group_id = GROUP_ID.fetch_add(1, Ordering::Relaxed);
+        let highest_group_id = HIGHEST_GROUP_ID.fetch_max(group_id, Ordering::AcqRel);
+
+        if group_id >= highest_group_id {
+            let group_id = NonZeroUsize::new(group_id).expect("bug: GROUP_ID overflowed");
+            Some(AllocationGroupId(group_id))
+        } else {
+            None
+        }
+    }
+
+    /// Attaches this allocation group to a [`Span`][tracing::Span].
+    ///
+    /// When the span is entered or exited, the allocation group will also transition from inactive to active, and vise
+    /// versa. In effect, all allocations that occur while the span is entered will be associated with the allocation
+    /// group.
+    pub fn attach_to_span(self, span: &Span) {
+        tracing::dispatcher::get_default(move |dispatch| {
+            if let Some(id) = span.id() {
+                if let Some(ctx) = dispatch.downcast_ref::<WithAllocationGroup>() {
+                    (ctx.with_allocation_group)(dispatch, &id, AllocationGroupToken::from(self));
+                }
+            }
+        });
+    }
+}
+
+/// A token that allows controlling when an allocation group is active or inactive.
+pub struct AllocationGroupToken {
+    id: AllocationGroupId,
+}
+
+impl AllocationGroupToken {
+    pub fn enter(&mut self) {
+        push_group_to_stack(self.id);
+    }
+
+    pub fn exit(&mut self) {
+        let current = pop_group_from_stack();
+        debug_assert_eq!(
+            current, self.id,
+            "popped group from stack but got unexpected group"
+        );
```

Comment:
Does this actually ever _not_ hit the assertion?

`GroupStack::pop` returns the _previous_ current allocation group, so if group A pushes to the stack, then group B pushes to the stack, and we exit group B, we'd get back group A as the result of `pop_group_from_stack`, I believe?